File: tree/Traversal/Inorder/recursive/java/InorderTraversal.java
import java.util.*;

public class InorderTraversal {
    public static List<Integer> InorderTraversal(int[] leftChildren, int[] rightChildren, int root) {
        List<Integer> order = new ArrayList<>();
        traverse(root, leftChildren, rightChildren, order);
        return order;
    }

    private static void traverse(int node, int[] leftChildren, int[] rightChildren, List<Integer> order) {
        int leftChild = leftChildren[node];
        if (leftChild != -1) {
            traverse(leftChild, leftChildren, rightChildren, order);
        }

        order.add(node);

        int rightChild = rightChildren[node];
        if (rightChild != -1) {
            traverse(rightChild, leftChildren, rightChildren, order);
        }
    }
}

File: tree/Traversal/Inorder/recursive/python/InorderTraversal.py
from collections.abc import Sequence


def InorderTraversal(left_children: Sequence[int], right_children: Sequence[int], root: int) -> list[int]:
    order: list[int] = []

    def traverse(node: int) -> None:
        left_child = left_children[node]
        if left_child != -1:
            traverse(left_child)

        order.append(node)

        right_child = right_children[node]
        if right_child != -1:
            traverse(right_child)

    traverse(root)
    return order
